FLIP: FCL Tx Payer Service (#612)
onflow/flow: flips/20210824-fcl-tx-payer-service.md

for their users. While this makes for a desireable user experience for those that use
Wallets with this feature, applications do not get the guarantee that _all_ of their
users will get this user experience. This is because users are able to use whichever FCL
compatible wallet they desire with the FCL compatible applications they use, and not
all FCL compatible wallets will strictly choose to provide this feature.

Applications that wish to guarantee that their users are able to submit transactions
without having to pay for them will then have to always pay for them themselves. These
applications would then have to build a "Transaction Fee Payer Service" that is able
to receive a transaction signable and then produce a signature for it.

We imagine that there is a potential business model here. A transaction fee payer
service (or: "TPaaS", "Transaction Payer as a Service") could be a webservice that
works with applications to provide payer signatures for the users of those applications.
Applications, instead of having to each individually build their own transaction
fee payer service, could depend on a this businesses implementation. The business could charge

> a fee to the applications that use it for their ability to use this service.

Author: I believe this specific version of the file you reviewed is outdated! You can view the latest version here: https://github.com/onflow/flow/pull/612/files

## User Benefit

Most consumer applications that we interact with on a daily basis charge us nothing.
As internet users, we're freely able to interact for our favourite social media apps,
interact with our favourite websites and simply do as we please without paying for anything.
However, with blockchain applications, each one of those interactions would likely be
transaction, and with that, transaction fees would have to be paid.

Users have been conditioned to expect things for free. As such, applications that wish
to provide the best user experiences will likely choose to allow their users to execute
transactions without them having to pay. 

This Transaction Fee Payer Service will enable more applications to be able to provide
this higher level of user experience. Creating value for the application, the users of
that application, and the potential for the tranaction fee payer service to capture some of that

> value as revenue.

Author: I believe this specific version of the file you reviewed is outdated! You can view the latest version here: https://github.com/onflow/flow/pull/612/files

# FCL Transaction Payer Service

| Status        | Proposed                                             |
:-------------- |:---------------------------------------------------- |
| **FLIP #**    | [612](https://github.com/onflow/flow/pull/612)       |
| **Author(s)** | Jeffrey Doyle (jeffrey.doyle@dapperlabs.com),        |
| **Sponsor**   |                                                      |
| **Updated**   | 2020-09-03                                           |

## Objective

> FCL provides a mechanism for a transactions fees to be paid by an account

Reviewer: @JeffreyDoyle 'FCL provides a mechanism for transactions fees to be paid by an account'?

Author: Removed this line!

to receive a transaction signable and then produce a signature for it.

A re-usable Transaction Payer Service would exist as project / set of packages and tools that applications could use
instead of having to build an implementation of them themselves. These re-usable components
would lower the overall complexity and effort required of a developer who wishes that 
their application provide "free" transactions for their users.

## User Benefit

Most consumer applications that we interact with on a daily basis charge us nothing.
As internet users, we're freely able to interact for our favourite social media apps,
interact with our favourite websites and simply do as we please without paying for anything.
However, with blockchain applications, each one of those interactions would likely be
transaction, and with that, transaction fees would have to be paid.

Users have been conditioned to expect things for free. As such, applications that wish
to provide the best user experiences will likely choose to allow their users to execute
transactions without them having to pay. 

This Transaction Fee Payer Service will enable more applications to be more easily able to provide

> a higher level of user experience.

Reviewer: 'This Transaction Fee Payer Service will enable more applications to offer a top of the line user experience, where users can easily and seamlessly engage in any action without friction.'

# FCL Transaction Payer Service

| Status        | Proposed                                             |
:-------------- |:---------------------------------------------------- |
| **FLIP #**    | [612](https://github.com/onflow/flow/pull/612)       |
| **Author(s)** | Jeffrey Doyle (jeffrey.doyle@dapperlabs.com),        |
| **Sponsor**   |                                                      |
| **Updated**   | 2020-09-09                                           |

## Objective

There is a need for the developement of a suite of tools that will collectively
act as a Transaction Fee Payer Service for use by applications built on Flow.

Technically, this suite of tools will facilitate signing as the payer of transactions initiated
by application's users with an account controlled by the application itself.

From a functional perspective, the service would enable users of applications that employ
these tools to be able to execute transactions, and thereby interact with Flow, without

> encurrign the burden of having to pay for their transaction fees themselves. This user experience

Reviewer: encurrign should be encurring

> is valuable for many reasons, including enabling users to engague with applications while not

Reviewer: engague should be engauge

having _any_ FLOW in their account, or having to worry about having _enough_ FLOW to execute the 
transactions they wish to execute.

A Transaction Fee Payer Service would exist as a resusable codebase / set of
packages & tools that could be used by the applications that desire it.

## Motivation

One of the many features of Flow is that transaction fees can be paid by an account
separate from the authorizers of that transaction. With this feature, applications that wish to 
provide "free" transactions for their users could opt to pay for their users' 
transactions themselves. They would do this by producing a signature for the 
transaction as the payer of that transaction. Upon receiving this transaction, 
the Flow network would then deduct transaction fees from the payer, not any of 
the authorizers.

For a user to pay their transaction fees themselves, that user would have to have FLOW
in their account. This means that should a first time Flow user attempt to engague with an application
that does not employ a Transaction Payer Service, they would first need to acquire FLOW before

> engaguing with that application. This is an undesireable user experience, as it significantly raises

Reviewer: 'This means that should a first time Flow user attempt to engague with an application
that does not employ a Transaction Payer Service, they would first need to acquire FLOW before
engaguing with that application' 
change to: 
'This can lead to a situation where a first-time Flow user attempts to engage with an application that does not employ a Transaction Payer Service, and they would first need to acquire FLOW before engaging with that application.'

> the friction required of first time Flow users to engague with these applications. Many users who are unfamiliar 

Reviewer: engague to engage

with blockchain applications will not understand or empathize with this requirement. Thus it is crucial
that applications who wish to provide a low friction experience for their users pay transaction fees on
their behalf.

Some FCL compatible wallets choose to act as the payer of the transactions their users
sign. They do this by swapping out their users account with an account of their own,
if their users are specified as the payer of a transaction. This allows wallets with
this feature to provide "free" transactions for their users. While this makes for a 
desirable user experience for those that use wallets with this feature, applications do not 
get the guarantee that _all_ of their users will get this user experience. 
This is because users are able to use whichever FCL compatible wallet they desire with the FCL 
compatible applications they use, and not all FCL compatible wallets will strictly choose to provide this feature.

> Applications that wish to guarantee their users can submit transactions
> without having to pay transaction fees for them will then have to pay those fees themselves. 

Reviewer 2: Slightly confusing... how about:
```suggestion
Applications that wish to guarantee their users can submit transactions
without having to pay transaction fees will have to pay those fees in the users' stead. 
```
